activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -533,8 +533,57 @@ def insert_fixtures(*)
         without_sql_mode("NO_AUTO_VALUE_ON_ZERO") { super }
       end
 
+      def insert_fixtures_set(fixture_set, tables_to_delete = [])
+        iterate_over_results = -> { while raw_connection.next_result; end; }
+
+        with_multi_statements do
+          without_sql_mode("NO_AUTO_VALUE_ON_ZERO") do
+            super(fixture_set, tables_to_delete, &iterate_over_results)
+          end
+        end
+      end
+
       private
 
+        def combine_multi_statements(fixture_inserts, tables_to_delete)
+          super
+
+          fixture_inserts.each_with_object([]) do |sql, total_sql|
+            previous_packet = total_sql.last
+            sql << ";\n"
+            if max_allowed_packet_reached?(sql, previous_packet) || total_sql.empty?
+              total_sql << sql
+            else
+              previous_packet << sql
+            end
+          end
+        end
+
+        def max_allowed_packet_reached?(current_packet, previous_packet)
+          if current_packet.bytesize > max_allowed_packet
+            raise ActiveRecordError, "Fixtures set is too large #{current_packet.bytesize}. Consider increasing the max_allowed_packet variable."
+          elsif previous_packet.nil?
+            false
+          else
+            (current_packet.bytesize + previous_packet.bytesize) > max_allowed_packet
```

Comment:
Do we need to allow for any padding etc here? "Packet" sounds to me like it could be including a surrounding struct of some sort. I just tried to check this experimentally, and ended up just confusing myself: in the `mysql` CLI tool, after setting `max_allowed_packet` to 1024, I was able to `SELECT 'xxxx'` for 10 KB of 'x's :confused:

Reply:
I got confused as well and should have explained beforehand.

> The packet message buffer is initialized to net_buffer_length bytes, but can grow up to max_allowed_packet bytes when needed. This value by default is small, to catch large (possibly incorrect) packets.

Because the [net-buffer-length](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_net_buffer_length) by default is set to 16384 bytes, setting the `max_allowed_packet` to 1024 has actually no effect. It will have effect only if `max_allowed_packet` is higher than the `net-buffer-length`, otherwise it doesn't need to shrink.

Reply:
Ah, thanks. After setting them both to 1024, I was able to run queries up to length **1022**, plus a trailing semicolon and/or newline (omitting them didn't allow me to add another character, so I'm not counting them). I haven't dug in to work out whether that's a quirk of the `mysql` CLI (e.g., maybe it always adds a semicolon if one isn't present?)

That's certainly close enough that it clearly doesn't count any surrounding data... but can you perhaps experiment with this real code, and work out whether we need to give it a 1-2 byte buffer? I guess it's just a matter of starting with a pair of fixtures whose inserts sum to ~1030, confirming they get sent as two executes, then removing one character at a time down to ~1020 (which should be sent as a single multi), and ensuring that each one of those runs succeeds.